### PR TITLE
Async restore

### PR DESCRIPTION
**Pull request by snopoke**
Currently incomplete but putting out for review / interest.

Spec: https://docs.google.com/document/d/1_kY3prhd3YSHzUbXLkfQQBenswyp6nJWCQL0IFC2WAs/edit#

@czue @benrudolph @ctsims

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/phone/data_providers/__init__.py [hunk 1]
```diff
@@ -1,5 +1,12 @@
-from .standard import *
-from .case import *
+from .standard import (
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/__init__.py]: F401 'LongRunningRestoreDataProvider' imported but unused

File: corehq/ex-submodules/casexml/apps/phone/data_providers/standard.py [hunk 2]
```diff
@@ -36,15 +36,30 @@ def get_elements(self, restore_state):
...
+
+class FixtureElementProvider(FixtureElementMixin, RestoreDataProvider):
+    pass
+
+
+class FixtureResponseProvider(FixtureElementMixin, LongRunningRestoreDataProvider):
+    def get_response(self, restore_state):
+        response = restore_state.restore_class()
+        response.extend(self.get_elements(restore_state))
+        return response
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/standard.py]: W292 no newline at end of file

Comment (ctsims) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/standard.py]: Awesome!

Let me know when this is up on staging so we can start running tests?

-Clayton

**Comment by benrudolph** [comment]
This is interesting. What if you thought about the restore as a bunch of transactions. on onset we create the list of transactions, in order, that need to happen (most likely some sort of call to `get_sync_element`). Once you have a list of transactions you continue to execute them until you run out of time. those transactions that were completed were marked as so. on retry, you only need to know where you left off in the list of transactions.

another tidbit that'd be cool, is that if these transactions were in order, we could process them in parallel and then finalize all the pieces into one file at the end.

obviously grazing over a lot of details, but just a few thoughts after looking at this

**Comment by snopoke** [comment]
Chatted with @czue about this and decided to put it on hold for now expecting that it won't be necessary with the new restore code.

**Comment by benrudolph** [comment]
this may yet be useful if we find that restores are causing bottlenecks in
cloudcare, similar to how touchcare-filter-cases is currently

**Comment by snopoke** [comment]
Can resurrect if we need to.

**Comment by ctsims** [comment]
I think we still want to do this, for what it's worth. The new restore
stuff doesn't work for domains with delegation and it's unbelievably
frustrating for our partners to deal with the restore lag.

@Ben: Wills work will possibly improve the touchforms stuff more than
anything else, assuming incremental restores are better than full restores?

-Clayton

**Comment by benrudolph** [comment]
definitely agree with incremental vs. full, but just pointing out that in
the event that we have a bunch of long running requests from restores that
we may want to think about making restore async. but that's pretty far in
the future, we still need to see how the new infrastructure performs

**Comment by ctsims** [comment]
I don't think we should drop this unless we can justify that we are not
actively getting regular timeouts on restores, which last I checked was
very true.

We can see what happens after the cleanliness push goes through, but I
dont' wanna keep punting this problem to the future.

-Clayton

**Comment by czue** [comment]
(we're not dropping just dropping priority for now until we think it's priority again)

**Comment by ctsims** [comment]
What's the level of effort for having a report on # of restore attempts
that result in

1) Success
2) 412 restore
3) Timeout
4) Error

that we can look at over time/for a period? My experience in the field has
been that this problem is very frequent, and we missed recently that phones
were 412'ing _way_ more often that we realized. It seems like we would make
better resourcing decisions if we had this info. Even if we expect this to
be fixed with owner cleanliness, would be good to see that verified in the
report, right?

-Clayton

**Comment by snopoke** [comment]
Completely agree that such a report would be useful. I'm not sure we can
know when there's a timeout though. @czue had been adding some more detail
to the synclog report so world be best to determine loe.

**Comment by czue** [comment]
the tricky thing is that this would have to come from the proxy to include timeouts.

I have the goal of having no 412s on the new restore code (though that is proving to be a challenge so far). 

I think let's revisit this in a month or so when I have a clearer picture of whether that goal will be feasible

**Comment by snopoke** [comment]
@czue @ctsims 
I was looking at the log data and did a quick analysis of the last 10000 or so OTA restores over the last 3 days or so. Our biggest errors are 499s (8.9%) which occur when the client disconnects.

image

@czue I'd be keen to do an experiment where we measure the stats for a domain prior to new sync being enabled and then again afterward so that we can do a comparison.

**Comment by ctsims** [comment]
And just to verify: The 401's are digest hits, so can be removed, right?
500's are Q errors, and 412's are expected (well, not "happy" expected, but
"Part of our sync process)

So of the remaining restores, which are restores we "expect" to succeed
we're at about a 19.5% failure rate on 499's. Not sure how easy it is for
us to parse through how much of that is due to client timeout v. client
connectivity woes, last time I looked through with Kibana the Timeouts were
easy to spot based on the connection time, not sure if we still have that
data.

-Clayton

**Comment by snopoke** [comment]
Unfortunately we don't get any data for the 499 durations at present, but
I'll play with the log format and see if we can't.

On 24 August 2015 at 23:00, ctsims <notifications@github.com> wrote:

> And just to verify: The 401's are digest hits, so can be removed, right?
> 500's are Q errors, and 412's are expected (well, not "happy" expected, but
> "Part of our sync process)
>
> So of the remaining restores, which are restores we "expect" to succeed
> we're at about a 19.5% failure rate on 499's. Not sure how easy it is for
> us to parse through how much of that is due to client timeout v. client
> connectivity woes, last time I looked through with Kibana the Timeouts were
> easy to spot based on the connection time, not sure if we still have that
> data.
>
> -Clayton
>
> On Mon, Aug 24, 2015 at 10:24 AM, Simon Kelly <notifications@github.com>
> wrote:
>
> > @czue <https://github.com/czue> @ctsims <https://github.com/ctsims>
> > I was looking at the log data and did a quick analysis of the last 10000
> > or so OTA restores over the last 3 days or so. Our biggest errors are
> 499s
> > (8.9%) which occur when the client disconnects.
> >
> > [image: image]
> > <
> https://cloud.githubusercontent.com/assets/249606/9442403/1c532b30-4a7c-11e5-8bab-ef86a52cb84c.png
> >
> >
> > @czue <https://github.com/czue> I'd be keen to do an experiment where we
> > measure the stats for a domain prior to new sync being enabled and then
> > again afterward so that we can do a comparison.
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/dimagi/commcare-hq/pull/6952#issuecomment-134223954
> >.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/6952#issuecomment-134378881>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

**Comment by czue** [comment]
@snopoke that sounds like a good experiment but i want to make sure i iron out the 412s first.

@ctsims i think the 401s are digest but not sure how to be 100% sure.

**Pull request closed without merging**